Bots - Shoot through glass + penetrateables

nullsystem opened the pull request:
<!--
Before submitting a pull request, ensure the following has been done:
* The branch has been tested with the latest master changes rebased in
* Fill in the descriptions, link the issues, and put in tags appropriate to the PR
* Update any documentation and comments if needed
* For WIP/Work in Progress PRs, use the Draft PR feature
-->

## Description
* Bots should now be able to shoot through glass if not using a shotgun
* If using a shotgun, try to change to a weapon that can utilize the glass if possible
* Change back to a shotgun if not behind glass?
* Tested/check if it doesn't straight up shoot through any penetrable like wallhack ability, non-glass penetrable only shoot through if it's recently seen enemy
* Enemy player that can shoot towards the ghoster behind penetrable should be able to try shooting at it
* Test/check for non-penetrable glass (tarmac non-penetrable vs oilstain penetrable)
* Slight scale between weapon penetration on when the bots will attempt to do penetration, so like SRS will do longer distance than SRM.
* Limited wallbang at ghoster ability to hard + expert bots

## Toolchain
<!--
If this is documentation only update, just remove the whole Toolchain section
NOTE: It's not needed for all to be filled in, just keep the toolchain/OS lines this PR been worked on

- Windows MSVC VS2022
-->
- Linux GCC Distro Native Gentoo/GCC 14

## Linked Issues
<!--
Applying issues here will auto-link the PR to its related issues if starting with "resolves".
If there's a related PR but don't want to resolve/close the issue, mark them with "related".

See: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword
-->
* fixes #1282

sunzenshen reviewed (dismissed):
LGTM, bots now can shoot through glass when they couldn't do so before.


https://github.com/user-attachments/assets/03acaac2-80b6-488f-84f8-b0e414067e9b



https://github.com/user-attachments/assets/367b67c2-1b21-4064-aa2d-e781d9af4bbb

Rainyan reviewed (approved):
> TODO (LATER): Test/check for non-penetrable glass?

Tested against the tarmac non-penetrable glass, and indeed it seems the bots will frutilessly empty their magazine against the bulletproof windows.

To repro:
```
neo_bot_quota 2;
map ntre_tarmac_ctg;

// Spawn to the garage side, so the bot spawns on the tarmac side.

// Then, teleport near the bulletproof window.
setang 7.062000 13.297527 0.000000;
setpos 477.418365 -252.499146 -210.468750;
```

But we can leave it for some future PR if you don't feel like it belongs here.

nullsystem commented:
@Rainyan I've added in glass penetration vs non-penetration check, just does an extra trace then penetration check without distance limitation.

@sunzenshen I've added in distance restriction scaling by weapon's penetration, so for example: a bot using an SRS will attempt a further distance than MX which will attempt further than SRM.

sunzenshen commented on src/game/server/neo/bot/neo_bot.cpp:
```diff
@@ -1693,50 +1701,173 @@ bool CNEOBot::IsQuietWeapon(CNEOBaseCombatWeapon* weapon) const
+	static constexpr const float FL_METERS_TRY_PENETRATE_MAX = 36.0f;
+
+	int material = physprops->GetSurfaceData(tr.surface.surfaceProps)->game.material;
+	if (material != 'J')
```
nit: maybe a comment explaining what 'J' refers to may be helpful.

Rainyan replied:
^ It's a bit hard to google for, so maybe adding a comment with link to https://developer.valvesoftware.com/wiki/Material_type wouldn't hurt.

Or well, having it refactored to some saner enumeration wouldn't hurt either but probably out of scope.

nullsystem replied:
It seems to not even been defined in `decals.h` properly, but assuming the original penetration code states `'J'` is blockbullets, I've `ifdef NEO` in `CHAR_TEX_BLOCKBULLETS` as `'J'` and used that define.

AdamTadeusz replied:
https://tenor.com/view/burnt-peanut-the-burnt-peanut-side-eye-gif-14097258958682501049

Rainyan commented:
> Bots appear to open fire on a player that is holding the ghost behind a paper wall near the house spawn on Isolation

Maybe you meant this already as a positive, but this is probably the desired behaviour, since good players should absolutely wallbang the ghoster when they have the chance.

It'd be really cool if we had some kind of "confidence level" that the bots had on the enemy location, and if they felt confident enough (seen player recently in area, sound cues, etc), they could outright start wallbanging thin surfaces to do preemptive damage, like a real high-level human player would. But that's out of scope for this PR surely.

Rainyan reviewed (changes requested):
For the tarmac bulletproof glass case mentioned before, it seems that if you are holding the ghost, the bots will still attempt to wallbang.

If it's not too much work, could you make it such that even if the bots have vision of the ghoster (possibly through a bullet-proof window), if they know their bullets cannot penetrate (the wall thickness check), they won't waste bullets trying to shoot through the impenetrable areas, but instead consider it as impossible and choose whatever is their alternative action to shooting instantly.

sunzenshen commented:
Ghost behind ballistic glass now works as expected:

https://github.com/user-attachments/assets/898fbb74-284a-4b34-84be-3716d42fb4ce
